### PR TITLE
 Server Side for Multi Battles (New)

### PR DESCRIPTION
**Pull request by Spandamn**
Old:   #7607 
Client: smogon/pokemon-showdown-client#1649
Last SS:
image

### CODE REVIEW

**Comment by Spandamn** [comment]
This should do it

**Comment by Spandamn** [comment]
oh yeah babyyy

**Comment by Spandamn** [comment]
halp Mac

File: sim/battle.ts [hunk 1]
```diff
@@ -1610,19 +1639,35 @@ export class Battle {
...
 		for (const side of this.sides) {
-			this.add('teamsize', side.id, side.pokemon.length);
+			let teamsize = side.pokemon.length;
+			if (format.teamLength && format.teamLength.battle) {
+				teamsize = format.teamLength.battle <= teamsize ? format.teamLength.battle : teamsize;
+			}
+			this.add('teamsize', side.id, teamsize);
 		}
 
-		this.add('gametype', this.gameType);
```
Comment (Zarel) [inline sim/battle.ts:1621]: Was this intentionally removed?

Comment (Zarel) [reply sim/battle.ts:1621]: Aha! It was moved to `room-battle`!

File: server/ladders.ts [hunk 2]
```diff
@@ -511,19 +563,37 @@ class Ladder extends LadderStore {
 			Ladders.searches.set(formatid, formatTable);
 		}
 		if (formatTable.has(user.id)) {
-			user.popup(`Couldn't search: You are already searching for a ${formatid} battle.`);
+			// user.popup(`Couldn't search: You are already searching for a ${formatid} battle.`);
```
Comment (Zarel) [inline server/ladders.ts:566]: Was this an oversight?

File: data/mods/gen7/random-teams.ts [hunk 3]
```diff
@@ -952,6 +952,18 @@ export class RandomGen7Teams extends RandomTeams {
 			(species.randomDoubleBattleMoves || species.randomBattleMoves) :
 			species.randomBattleMoves;
 		const movePool = (randMoves || Object.keys(Dex.getLearnsetData(species.id).learnset!)).slice();
+		if (this.format.gameType === 'multi') {
+			// Random Multi Battle uses doubles move pools, but Ally Switch fails in multi battles
+			const allySwitch = movePool.indexOf('allyswitch');
+			if (allySwitch !== undefined) {
```
Comment (AnnikaCodes) [inline data/mods/gen7/random-teams.ts:958]: This doesn't work. [`Array#indexOf`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf) never returns `undefined`; if the value is not in the array, it returns -1. I fixed this in 2baca9339d52e9523dd0dd9f9105eb9309743ff0.

File: sim/battle.ts [hunk 4]
```diff
@@ -1524,15 +1546,17 @@ export class Battle {
...
 		}
 
+		this.add('gametype', this.gameType);
 		for (const side of this.sides) {
-			this.add('teamsize', side.id, side.pokemon.length);
+			let teamsize = side.pokemon.length;
+			if (format.teamLength && format.teamLength.battle) {
+				teamsize = format.teamLength.battle <= teamsize ? format.teamLength.battle : teamsize;
+			}
+			this.add('teamsize', side.id, teamsize);
```
Comment (Zarel) [inline sim/battle.ts:1555]: Wait wtf what does this mean? Why was this done?

File: sim/battle.ts [hunk 5]
```diff
@@ -2040,26 +2078,20 @@ export class Battle {
...
-				const foeActives = pokemon.side.foe.active;
-				const frontPosition = foeActives.length - 1 - pokemon.position;
-				let adjacentFoes = foeActives.slice(frontPosition < 1 ? 0 : frontPosition - 1, frontPosition + 2);
-				adjacentFoes = adjacentFoes.filter(active => active && !active.fainted);
+				const adjacentFoes = pokemon.adjacentFoes();
 				if (adjacentFoes.length) return this.sample(adjacentFoes);
-				// no valid target at all, return a foe for any possible redirection
-				return foeActives[frontPosition];
+				// no valid target at all, return a possibly-fainted foe for any possible redirection
+				return pokemon.side.foe.active[0];
```
Comment (urkerab) [inline sim/battle.ts:2094]: This change broke redirection in Triples.

**Pull request merged**